### ISSUE TITLE
syntactic part of N999 in another E rule

### ISSUE THREAD

**Issue by edhinard** [astral-sh/ruff#12475]
Not only does rule N999 check that the PEP8 convention is respected, but also checks that the module name is syntactically correct. The latter is not part of PEP8. This is a problem for me currently working on "old non linted" code. I have to ignore this rule because of some modules whose names are syntactically correct but sometimes with capital letters.
I think a E rule is missing that would only include the syntax checking of N999. (If I were you I would also remove that part from N999 for consistency)

**Comment by MichaReiser** [comment]
It doesn't seem that Python requires all module names to be valid identifiers. Only modules that are imported from other modules. But I might be wrong here. Ruff should always report a syntax error if you try to import a module with an invalid identifier name.

**Comment by AlexWaygood** [comment]
I believe @edhinard is complaining about something like the following directory structure:

- `foo/`
  - `foo/__init__.py` (empty file)
  - `foo/NotPep8Compliant.py`
  - `foo/not-an-identifier.py`

They'd like a rule that complains about the naming of `foo/not-an-identifier.py`, as importing `foo.not-an-identifier` at runtime will fail. But they can't use our [`N999`](https://docs.astral.sh/ruff/rules/invalid-module-name/) rule (`invalid-module-name`) to flag this, because that would also flag `foo/NotPep8Compliant.py`). None of the other submodules in the `foo` package ever try to import `not-an-identifier` -- it's just exposed as a public submodule of the `foo` package -- so the bad name of the submodule also isn't caught by any of our rules that flag `SyntaxError`s.

So I think the argument is that N999 should be split into two rules: one that flags invalid module names, and one that flags module names that are valid, but not PEP8-compliant. This seems like a reasonable argument to me, though it would mean we'd deviate from the behaviour of the original pep8-naming linter.

**Comment by edhinard** [comment]
You are right module names don't have to be valid identifiers:
```shell
$ touch 3.py
$ python3 -c "import importlib; print(importlib.import_module('3'))"
<module '3' from '/tmp/3.py'>
```
But it is not a good idea:
```shell
$ python3 -c "import 3"
  File "<string>", line 1
    import 3
           ^
SyntaxError: invalid syntax
```
and ruff is here to help us avoid that.
IMHO the problem with N999 is that it combine two checks: 1) the pep8 convention which we cannot always adhere to, 2) the "dangerous permissiveness" of non regular module names. It is important to have 2) - for example sub-directories may be implicitely imported by pytest.
Probably a new "E" rule is not a good idea. But something not mixed with pep8 convention

**Comment by MichaReiser** [comment]
@AlexWaygood would you support this check as a separate rule?

**Comment by AlexWaygood** [comment]
> @AlexWaygood would you support this check as a separate rule? @AlexElvers is interested in it.

Are you asking if I'd support splitting the existing check up (as the OP in this issue is requesting) or if I'd support adding an additional, more opinionated rule that also checked N999-type things for files that are not part of packages (as was proposed in https://github.com/astral-sh/ruff/pull/12535)?

For the first question: yes, definitely. For the second question: I think it could be a useful rule for some users, but I think it would probably be too opinionated for us to add now until we've done rule recategorisation.

**Comment by MichaReiser** [comment]
Oh, I thought it's both the same. Thanks for explaining and makes sense to me

**Comment by bhattiishfaq804-max** [comment]
I like it

**Comment by bhattiishfaq804-max** [comment]
I love it

**Comment by bhattiishfaq804-max** [comment]
Nice one I

**Comment by bhattiishfaq804-max** [comment]
Love you

**Comment by Avasam** [comment]
With the new set of default configuration, I'd like to also push for spitting this rule in two:
The default configs only includes `N999` from the `pep8-naming` family of rules. This tells me that the goal was not to enforce any arbitrary naming convention here, only to catch modules that can't be imported regularly (which makes more sense for a default set of rules).

**Comment by ntBre** [comment]
Looking back at my notes, I did categorize it as a Style rule (the lowest severity level that we included in the defaults), but I can't find any specific discussion of treating it differently from the other pep8-naming rules, so I think you're probably right. The cases that can't be imported probably pushed it up to Style, whereas the naming convention part could drop to Pedantic. Most of the other PEP 8 rules I checked also ended up in Pedantic, so that seems consistent.

In short, I'd be supportive of this split.

**Comment by jakobtfr** [comment]
Hi, I would like to work on this.

Reading the discussion, my approach would be:
  - Keep N999 for invalid module names that cannot be used in regular import statements.
  - Add a separate PEP 8 naming rule.
Questions: Is N998 the preferred code? Should ignore-names apply only to the style rule?

**Comment by Avasam** [comment]
`N9XX` doesn't feel right given `MyModule.py` is completely valid identifier. Conventionally in flake8 the 9XX codes are used for straight up errors, or codes you wouldn't enable by default.
https://flake8.pycqa.org/en/latest/user/error-codes.html#error-violation-codes:~:text=E999
https://pycodestyle.pycqa.org/en/latest/intro.html#error-codes:~:text=E9

Even outside of that, I wouldn't group "invalid" module name with "unconventional" module name (the first digit of a code acting like a group for selection, `N9` vs `N8`)

This also reveals a bit of a misnomer with the first few `N8` rules, where the name says "invalid", whilst it's more so "unconventional". But that may be getting off-topic.

**Comment by ntBre** [comment]
I don't really know what the upstream naming/coding conventions are, but given the ongoing work on both #1773 and #1774, I'm not feeling too worried about specific categories or codes at the moment. N998 seems okay to me. RUF is another common catch-all category, but I wouldn't want to put the more pedantic variant of N999 in there.

Usually I'd suggest adding one to the previous rule code, but N1000 also seems undesirable.

It seems fine to let `ignore-names` apply to both rules. If anything, it probably shouldn't apply to N999 anymore because it doesn't really make sense to ignore an invalid name like `some-module.py`. Actually it's not really clear to me why `ignore-names` is needed for either rule. The rules apply to the filenames, so `per-file-ignores` would work just as well and also accepts globs. But preserving the current behavior in both seems okay for now.
